File: commands/leaderboard.py
import discord
from discord.ext import commands
from db_model import get_leaderboard


@commands.command(aliases=['lb'])
async def leaderboard(context, emoji: discord.Emoji):
    rows = get_leaderboard(context, emoji)

    try:
        embed = discord.Embed(
            colour=discord.Colour.orange()
        )
        embed.set_author(name=emoji.name + ' Leaderboard')
        embed.set_thumbnail(url=emoji.url)
        position = 1
        for row in rows:
            # print(row[0])
            try:
                embed.add_field(name=str(position) + ". " + str(context.bot.get_user(int(row[0]))), value=str(row[1]),
                                inline=False)
                position += 1
            except ValueError:
                # print('Value Error')
                embed.add_field(name=str(position) + ". " + str(row[0]),
                                value=str(row[1]),
                                inline=False)
                position += 1

        await context.send(embed=embed)
    except IndexError as error:
        # Catch for invalid output, outputs an embed to chat.
        print(error)
        embed = discord.Embed(
            colour=discord.Colour.orange()
        )
        embed.set_author(name='Nothing To Display!')
        embed.add_field(name='User input has no valid output.',
                        value='Possible reasons:'
                              '\n- Empty database'
                              '\n- No emoji-usage records at specified date range'
                              '\n- No emoji-usage records by specified user'
                        )
        await context.send(embed=embed)


@leaderboard.error
async def leaderboard_error(context, error):
    # print(error)
    embed = discord.Embed(
        colour=discord.Colour.orange()
    )
    embed.set_author(name='Improper Command Usage.')
    embed.add_field(name='Usage:', value='```ES leaderboard <emoji>'
                                         '\nES lb <emoji>```', inline=True)
    embed.add_field(name='Examples:',
                    value='```ES leaderboard :thonk:\nES lb :thonk:```',
                    inline=False)
    await context.send(embed=embed)


def setup(bot):
    bot.add_command(leaderboard)

File: commands/leaderboard.js
const { getLeaderboard } = require('../db_model');
const { MessageEmbed } = require('discord.js');
const { SlashCommandBuilder } = require('@discordjs/builders');

module.exports = {
	data: new SlashCommandBuilder()
		.setName('leaderboard')
		.setDescription('Displays the top ten users for a specified emote\'s usage')
		.addStringOption(option =>
			option.setName('type')
				.setDescription('The type of leaderboard to display.')
				.setRequired(true)
				.addChoices([
					['Sent', 'sent'],
					['Received', 'received'],
				]))
		.addStringOption(option =>
			option.setName('emoji')
				.setDescription('The emoji to get the leaderboard for.')
				.setRequired(true))
		.addIntegerOption(option =>
			option.setName('daterange')
				.setDescription('The date range to query for.')
				.addChoices([
					['All Time', 0],
					['Yearly', 365],
					['Monthly', 30],
					['Weekly', 7],
					['Daily', 1],
					['Hourly', 60],
				])),
	async execute(interaction) {
		const embed = new MessageEmbed().setColor('ORANGE');

		// Validate choices.
		let dateRange = interaction.options.getInteger('daterange');
		switch (dateRange) {
		case 0:
			// all time
			embed.setDescription('All-Time');
			dateRange = '0';
			break;
		case 365:
			// yearly
			embed.setDescription('Yearly');
			dateRange = new Date();
			dateRange.setDate(dateRange.getDate() - 365);
			dateRange = dateRange.toISOString();
			break;
		case 30:
			// monthly
			embed.setDescription('Monthly');
			dateRange = new Date();
			dateRange.setMonth(dateRange.getMonth() - 1);
			dateRange = dateRange.toISOString();
			break;
		case 7:
			// weekly
			embed.setDescription('Weekly');
			dateRange = new Date();
			dateRange.setDate(dateRange.getDate() - 7);
			dateRange = dateRange.toISOString();
			break;
		case 1:
			// daily
			embed.setDescription('Daily');
			dateRange = new Date();
			dateRange.setDate(dateRange.getDate() - 1);
			dateRange = dateRange.toISOString();
			break;
		case 60:
			// daily
			embed.setDescription('Hourly');
			dateRange = new Date();
			dateRange.setHours(dateRange.getHours() - 1);
			dateRange = dateRange.toISOString();
			break;
		default:
			embed.setDescription('All-Time');
			dateRange = null;
		}

		// Validates emoji option.
		const stringEmoji = interaction.options.getString('emoji');
		const re = /(?<=:)\d*(?=>)/g;
		const emojiIds = stringEmoji.match(re);
		let emoji = null;
		try {
			emoji = await interaction.guild.emojis.fetch(emojiIds[0]);
		}
		catch {
			return interaction.reply({ content: 'No emoji found in string.', ephemeral: true });
		}

		// Grabs leaderboard info.
		const type = interaction.options.getString('type');
		const array = (dateRange ?
			getLeaderboard(interaction.guild.id, emoji.id, interaction.client.id, type, dateRange) :
			getLeaderboard(interaction.guild.id, emoji.id, interaction.client.id, type));

		// Catch for empty leaderboard.
		if (!array.length) {
			return interaction.reply({ content: 'Sorry, there\'s no info to display!' });
		}

		// Fills embed.
		embed.setTitle(`${emoji.name} Leaderboard`).setThumbnail(`${emoji.url}`);
		let pos = 1;
		for await (const row of array) {
			const count = Object.values(row)[1];
			const userId = Object.values(row)[0];
			try {
				const user = await interaction.guild.members.fetch(userId);
				embed.addField(`${pos}. ${user.displayName}`, `${count}`);
			}
			catch (e) {
				console.error(e);
			}
			pos++;
		}
		return interaction.reply({ embeds: [embed] });
	},
};
